subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java

```diff
@@ -281,15 +292,89 @@ public void execute(GenerateEclipseClasspath task) {
                         }
                     }
                 });
+
+                SourceSetContainer sourceSets = project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets();
+                sourceSets.configureEach(new Action<SourceSet>() {
+                    @Override
+                    public void execute(SourceSet sourceSet) {
+                        if (sourceSet.getName().toLowerCase(Locale.ROOT).contains("test")) {
+                            // source sets with 'test' in their name are marked as test on the Eclipse classpath
+                            model.getClasspath().getTestSourceSets().convention(ImmutableSet.<SourceSet>builder()
```

Comment:
So each time we configure a new SourceSet, we build a new ImmutableSet containing all the existing SourceSets + the new one?  

Wouldn't it be more efficient to set the convention to a local mutable set declared outisde of `configureEach()`, and then have configure callback insert each new sourceset into the mutable set?

Reply:
Does this pattern not work?  If not, this block of logic to create a new immutable set by adding the new item to the existing set would be better extracted to a diferent private method.